### DISCUSSION TITLE
Add control via MQTT

### DISCUSSION THREAD

**Discussion by michmill1970** [teslamate-org/teslamate#1657] [Ideas]
I've seen a few threads about using TeslaMate to control vehicle features by sending MQTT messages to TeslaMate.  Is anyone working on this?  If not, I'll pick it up and see if I can get something working.  If there is already a branch, please let me know and I'll help out.

The goal for me is being able to control my Tesla via Home Assistant.  Unfortunately, the native Tesla integration for Home Assistant is broken, which seems to be a common issue.  My preference is to have only one place that integrates with the Tesla API, and TeslaMate seems like the best, logical choice.

**Comment by jun3280net** [comment]
> I've seen a few threads about using TeslaMate to control vehicle features by sending MQTT messages to TeslaMate. Is anyone working on this? If not, I'll pick it up and see if I can get something working. If there is already a branch, please let me know and I'll help out.
> 
> The goal for me is being able to control my Tesla via Home Assistant. Unfortunately, the native Tesla integration for Home Assistant is broken, which seems to be a common issue. My preference is to have only one place that integrates with the Tesla API, and TeslaMate seems like the best, logical choice.

@michmill1970  just wanted to let you know you can fix the Tesla integrations using the following link : 
[https://github.com/alandtse/tesla](https://github.com/alandtse/tesla)

Cheers

**Comment by adriankumpf** [comment]
TeslaMate is read-only by design. We don't want to send any commands to the Tesla API.

I think it would also be preferable to send commands via an HTTP API than via MQTT. There is an excellent TeslaMate extension which allows you to do just that (among other things): https://github.com/tobiasehlert/TeslaMateApi#commands

**Reply by LelandSindt** [reply]
Previous treads on this topic.. 
https://github.com/adriankumpf/teslamate/issues/601#issuecomment-616845343
https://github.com/adriankumpf/teslamate/issues/950

For what its worth, I do see some value in an asynchronous command queue, as long as failure without feedback is acceptable. 

If you are still inclined to build an MQTT based control/command service, I would encourage you build it to depend on [TeslaMateAPI](https://github.com/tobiasehlert/TeslaMateApi#commands) which in turn depends on TeslaMate to maintain the [auth tokens](https://github.com/tobiasehlert/teslamateapi/blob/main/src/v1_TeslaMateAPICarsCommand.go#L78)...

**Reply by LelandSindt** [reply]
@tobiasehlert  --- tagging you in case you have further thoughts on this subject...

**Reply by tobiasehlert** [reply]
Sound exactly like @michmill1970 should try [TeslaMateApi](https://github.com/tobiasehlert/TeslaMateApi#commands) :]

**Reply by michmill1970** [reply]
Sounds like I have some coding to do.  :)

I would appreciate feedback on my design plan:
To have a "unified" MQTT API, I'm going to listen for MQTT messages from TeslaMate.  I will republish those message with a slightly modified state topic and payload, and listen on the same modified topic.  Command messages received to my program via MQTT will be proxied into TeslaMateAPI REST command calls.
The goal is for both state and command messages to share both topics and general format.  This should allow easy integration into other applications like Home Assistant. I'm aiming for a light, stateless translation from TeslaMate MQTT, and to TeslaMateAPI REST.

Thoughts?

**Reply by LelandSindt** [reply]
Generally I am not a fan of re-publishing things. 

Reviewing the [list of topics](https://docs.teslamate.org/docs/integrations/mqtt) the majority are going to be read only.

for those that you can effect, I would be inclined to have sister topics... 

`teslamate/cars/$car_id/locked` -> `commandQueue/cars/$car_id/locks` and watch for `lock` or `unlock`.

That said, I know nothing about your HA use case, perhaps there are limitations to reading from one topic for state and writing to another topic for commands.. :shrug:

**Reply by LelandSindt** [reply]
What Language are you planning to use?

**Reply by michmill1970** [reply]
@LelandSindt, 
Thanks for the feedback.  I agree that normally I wouldn't republish, but Home Assistant only allows 1 topic for a state/command set of MQTT messages.  You can specify different templates for the state and command messages, but only a single topic for a given item (like a lock).  Perhaps I'll just republish items that have corresponding command messages, but as anyone in my development group will tell you I have a thing for consistency.

I haven't decided on a language yet.  I would normally do something like this in Python, but due to the simple nature of the design I think I'm going to use this to learn Go, or maybe Elixir.

**Reply by LelandSindt** [reply]
Sounds like a reasonable path. +1 for only republishing items that have corresponding command messages.

I am a big fan of Go and the [paho mqtt](https://github.com/eclipse/paho.mqtt.golang) library is well maintained and supported. I find that Python is fine, until you want to distribute it... (though Docker makes distribution easier...) and Elixir is dark magic...  :wink: (not that you asked my opinion.)

**Comment by sillyfrog** [comment]
@michmill1970 Have you had much luck with this? I'm considering something similar; using MQTT to make setting changes to my Tesla Car via HA.

I have published what I'm currently using here https://github.com/sillyfrog/tesla-mqtt - it's very rough around the edges and does not do the nice stuff that teslamate does to allow the car to go to sleep etc.

To clarify on an earlier comment on the implementation required for HA, you can have a separate _state_ topic vs _command_ topic, and for my implementation I have done something similar to the [Homie](https://homieiot.github.io/) convention and add `/set` to the end of a topic to allow changing a setting. For example, the current charge limit will be published to `tesla/car/charge-limit`, and you can set it at `tesla/car/charge-limit/set`.

In my current implementation, it will fail if the car is asleep (causing the car to start to wake up), but the current state will still be published, so my scripts know to retry. I'm using it to only charge my car when there is "spare" solar power and/or cheap grid power - it checks every five minutes, and adjusts the charge limit accordingly (eg: lots of spare power, charge to 80%, no spare power, only charge to 50% and everything inbetween) - that way if a message is missed, it'll get picked up again within 5 minutes, and the car is awake. That said, if it were to have an internal queue and retry, that should also be fairly simple to implement.

The ideal solution for me is something that's running within TeslaMate - but I don't know Elixir (however I'm considering giving it a shot in a fork, ideally written in a way that the fork can always merge in upstream easily as things are improved).

**Comment by sillyfrog** [comment]
I have reworked my implementation (mentioned above) to support Home Assistant, including MQTT Auto Discovery, MQTT message aggregation (for GPS updates), and commands for adjusting charge level and charge status. Check it out here: https://github.com/sillyfrog/teslabuddy

**Comment by pmboothby** [comment]
You can also use the Tesla API via Node-RED. I'm currently writing an integration guide for Teslamate->Node-RED. It's quite useful for this kind of integration.

**Comment by pmboothby** [comment]
BTW, here's a draft of my Teslamate/Node-RED integration guide:

https://github.com/pmboothby/teslamate/blob/master/website/docs/integrations/Node-RED.md

I've submitted a pull request, but if anyone wants to try this out and give me feedback on the documentation, that would be handy. 

Though not included in this guide, there is a Tesla Node-RED module that can send commands to the car, so you could program logic to perform some event like turn on the climate control based on a calendar or timer, etc.

**Comment by dranreb38** [comment]
@pmboothby Thanks for your project, that's what I'm looking for, for I am now on github ! 

![Screenshot_236](https://user-images.githubusercontent.com/16446911/139600153-0fe4c529-6af2-4b73-8f09-a59efccf3423.png)
:)

**Comment by dranreb38** [comment]
Hello @pmboothby ,
I set up your integration in node-red everything works correctly !, now I would like to receive a telegram message as soon as a door opens, I tried to add a function at the output of the switch but this does not work, need help to receive a message when doors_open is **TRUE**  
Thank

**Reply by pmboothby** [reply]
In the _Notifcations->state changes_ node, add something like this after the line `switch(msg.topic){`:
```
    case 'doors_open':
        msg.topic = `Doors open: * ${msg.payload} *`;
        break;
```
Here's output from a little test, getting into car, open/close door, get out:
![image](https://user-images.githubusercontent.com/34880264/141063448-48670cdc-1279-4820-a6d9-4131d2026181.png)

**Reply by dranreb38** [reply]
Hello @pmboothby , 
It's good ! , i use with node-red-contrib-sms-free-mobile, but the variables are no longer good! (msg.payload) empty after telegram. I'm looking for a solution
Thank

**Reply by pmboothby** [reply]
Hi @dranreb38. I am not sure that I understand what you are doing, but it sounds like you are connecting the output from the Telegram node to a node-red-contrib-sms-free-mobile node. What you can do is simply to connect the _Format message_ node to the SMS node, too. Here is an example where the connection goes to both the Telegram node and to an email node:
![image](https://user-images.githubusercontent.com/34880264/142828835-a97f67ea-0524-47cb-81ec-c858f71aa534.png)
Of course, you can replace the Telegram node if you just want to send SMS. Does this answer your question?

**Reply by pmboothby** [reply]
@dranreb38 I have noticed that if the door is opened and closed for only a short time that no messages are received. I don't know if this is the car, Tesla's API or Teslamate.

**Reply by dranreb38** [reply]
I solved my problem, I must send the msg.payload before entering the function "msg.payload.options = {parse_mode:" MarkdownV2 "}
![Screenshot_317](https://user-images.githubusercontent.com/16446911/142837031-7cb162be-58a2-4fc4-93d3-f2965ed27d78.png)

I don't understand this command msg.payload.options = {parse_mode: "MarkdownV2"}

**Comment by dranreb38** [comment]
Hello @pmboothby 
I would like to create a new function that allows sending a message when a door is open. I can't handle the condition doors_open is true, with if then else for example if doors_open == true send payload "Door opening" else "Door closed" **BUT** with only one msg per event, thanks for your help

**Comment by dranreb38** [comment]
I solved my problem, here is the flow to use. I just have to include "moment" for date time management in the message

![Screenshot_340](https://user-images.githubusercontent.com/16446911/145539456-b8212ec5-ee67-4920-a4f5-d1c78e1461c4.png)
.
